File: language-analysis/2.2.Performance-Memory-Analysis/linear_search.py
import sys
import timeit


data_list = [(i, f"item_{i}") for i in range(1000)]


def search_in_list(target):
    for key, value in data_list:
        if key == target:
            return value
    return None


target = 999

print("Implementation: Linear Search in a List")
print("Search result:", search_in_list(target))

execution_time = timeit.timeit(
    "search_in_list(target)",
    globals=globals(),
    number=100000,
)

memory_used = sys.getsizeof(data_list)

print("\nMetric 1: Execution Time")
print(f"{execution_time:.6f} seconds")

print("\nMetric 2: Memory Footprint")
print(f"{memory_used:,} bytes")

File: language-analysis/2.2.Performance-Memory-Analysis/linear_search.js
const { performance } = require("node:perf_hooks");

const dataList = Array.from({ length: 1000 }, (_, i) => [i, `item_${i}`]);

function searchInList(target) {
  for (const [key, value] of dataList) {
    if (key === target) {
      return value;
    }
  }
  return undefined;
}

function measureExecutionTime(iterations, target) {
  const start = performance.now();

  for (let i = 0; i < iterations; i += 1) {
    searchInList(target);
  }

  const end = performance.now();
  return (end - start) / 1000;
}

function estimateMemoryFootprint() {
  return Buffer.byteLength(JSON.stringify(dataList));
}

const target = 999;
const executionTime = measureExecutionTime(100000, target);
const memoryUsed = estimateMemoryFootprint();

console.log("Implementation: Linear Search in an Array");
console.log("Search result:", searchInList(target));

console.log("\nMetric 1: Execution Time");
console.log(`${executionTime.toFixed(6)} seconds`);

console.log("\nMetric 2: Memory Footprint");
console.log(`${memoryUsed.toLocaleString()} bytes`);
